### PR TITLE
Talk to nodes directly if connected via the reciever and add wired usb support.

### LINKED ISSUE #589: Talk to the device hidraw nodes directly
The `hid-logitech-dj` kernel driver exports a different hidraw node for each device connected to a receiver. We would like to make use of that and talk to the device directly instead of going trough the hidraw node of the receiver.

### PR DESCRIPTION
**Pull request by effective-light**
Fixes #589
Addresses #268 

Requires installing the new udev rule.

### CODE REVIEW

**Comment by pfps** [comment]
Wow!   This is a significant addition, but something that is quite useful.   It is going to require some careful reviewing.  It's too bad that your major changes didn't go through the style checks.

File: lib/logitech_receiver/base_usb.py [hunk 1]
```diff
@@ -128,7 +134,16 @@
 LIGHTSPEED_RECEIVER_C53f = _lightspeed_receiver(0xc53f)
 LIGHTSPEED_RECEIVER_C53d = _lightspeed_receiver(0xc53d)
 
-del _DRIVER, _unifying_receiver, _nano_receiver, _lenovo_receiver, _lightspeed_receiver
+# Wired devices
```
Comment (pfps) [inline lib/logitech_receiver/base_usb.py:133]: I think it would be more useful if there were descriptions of what these devices are.

File: lib/solaar/cli/show.py [hunk 2]
```diff
@@ -247,11 +247,11 @@ def run(devices, args, find_receiver, find_device):
                 dev_num += 1
         return
 
-    dev = find_receiver(receivers, device_name)
+    dev = find_receiver(devices, device_name)
```
Comment (pfps) [inline lib/solaar/cli/show.py:250]: Can these changes be amends to previous commits so that the commit history looks better?

File: lib/logitech_receiver/base.py [hunk 3]
```diff
@@ -95,11 +95,13 @@ def receivers():
         for d in _hid.enumerate(receiver_usb_id):
             yield d
```
Comment (pfps) [inline lib/logitech_receiver/base.py:97]: It would be nice to have this style change not be a separate commit but instead be part of the underlying commit.

File: lib/logitech_receiver/receiver.py [hunk 4]
```diff
@@ -133,6 +137,12 @@ def __init__(self, receiver, number, link_notification=None):
         # also it gets set to None on this object when the device is unpaired
         assert self.wpid is not None, 'failed to read wpid: device %d of %s' % (number, receiver)
 
+        for dev in _hid.enumerate({'vendor_id': 0x046d}):
```
Comment (pfps) [inline lib/logitech_receiver/receiver.py]: This fails on my system because the device hidraw nodes are not accessible by regular users.   Solaar has a udev rule that makes the receiver nodes accessible but not the device nodes.   
@FFY00 Do you have any ideas on how the udev rule could be extended to make these nodes accessible?

**Comment by pfps** [comment]
When I run as root (to get around the hidraw access problem) I'm getting an error in base.py.
```
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/listener.py", line 193, in run
    self._notifications_callback(n)
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/listener.py", line 213, in _notifications_handler
    dev = self.receiver.register_new_device(n.devnumber, n)
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/receiver.py", line 148, in register_new_device
    dev = Device(self, number, notification)
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py", line 132, in __init__
    print("DEVICE ENUMERATE", self, dev)
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py", line 356, in __str__
    self.number, self.wpid or self.product_id, self.name or self.codename or '?', self.serial
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py", line 197, in name
    if self.online and self.protocol >= 2.0:
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py", line 172, in protocol
    self._protocol = _base.ping(self.handle, self.number)
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/base.py", line 468, in ping
    ihandle = int(handle)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

File: lib/logitech_receiver/receiver.py (hunk 4 above)
Comment (pfps) [reply lib/logitech_receiver/receiver.py]: I *think* I have a solution, and it is simple.  Just remove this part of the udev rule:
```
# don't apply to the paired peripherals, just the receivers
DRIVERS=="logitech-djdevice|logitech-hidpp-device", GOTO="solaar_end"

```

Comment (pfps) [reply lib/logitech_receiver/receiver.py]: The wired devices will have to be added to the rule, though, but I think that is fine.

**Comment by pfps** [comment]
There is a bug in Solaar having to do with logging DJ repairing notifications.   I just fixed it but it was causing my testing of these changes to have some glitches.   You should rebase your commits on the current master branch.

**Comment by pfps** [comment]
When I pair my mouse (just after unpairing it) I see the following errors:
```
05:38:03,293    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (16) => r[10 02 4104 525140]
05:38:03,307    ERROR [ReceiverListener:hidraw0] logitech_receiver.listener: processing Notification(2,41,04,525140)
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/listener.py", line 193, in run
    self._notifications_callback(n)
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/listener.py", line 213, in _notifications_handler
    dev = self.receiver.register_new_device(n.devnumber, n)
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/receiver.py", line 148, in register_new_device
    dev = Device(self, number, notification)
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py", line 137, in __init__
    assert self.handle
AssertionError
05:38:03,308    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (16) => r[20 02 4201 0000000000000000000000]
05:38:03,309    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (16) <= w[10 FF 83B5 210000]
05:38:03,309    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (16) => r[20 02 4200 0000000000000000000000]
05:38:03,311    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (16) => r[11 02 0E00 01010100000000000000000000000000]
05:38:03,313    DEBUG [ReceiverListener:hidraw0] logitech_receiver.base: (16) => r[11 FF 83B5 21590840510402020C00000000000000]
05:38:03,319     INFO [ReceiverListener:hidraw0] logitech_receiver.receiver: <UnifyingReceiver(/dev/hidraw0,16)>: found new device 2 (4051)
05:38:03,319    ERROR [ReceiverListener:hidraw0] logitech_receiver.listener: processing Notification(2,42,01,0000000000000000000000)
Traceback (most recent call last):
  File "/home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/listener.py", line 193, in run
    self._notifications_callback(n)
  File "/home/local/SoftwareDownloads/Solaar/lib/solaar/listener.py", line 241, in _notifications_handler
    assert dev.status is not None
AttributeError: 'Device' object has no attribute 'status'
05:38:03,320     INFO [ReceiverListener:hidraw0] solaar.listener: ignoring DJ pairing notification Notification(2,41,00,5140040000000000000000)
05:38:03,320     INFO [ReceiverListener:hidraw0] logitech_receiver.notifications: <UnifyingReceiver(/dev/hidraw0,16)>: pairing lock is closed
05:38:03,320     INFO [ReceiverListener:hidraw0] solaar.listener: status_changed <UnifyingReceiver(/dev/hidraw0,16)>: present, 2 paired devices. (1) pairing lock is closed
05:38:03,320    DEBUG [MainThread] solaar.ui: status changed: <UnifyingReceiver(/dev/hidraw0,16)> (NOTIFICATION) pairing lock is closed
05:38:03,321    DEBUG [MainThread] logitech_receiver.base: (19) <= w[10 FF 8102 000000]
05:38:03,323    DEBUG [MainThread] logitech_receiver.base: (19) 
```

Also, Solaar doesn't notice that a device paired and leaves the pairing lock open until the timeout.   But in the end my mouse was paired.

**Comment by pfps** [comment]
I was looking into what it would take to add Bluetooth support to this and it doesn't appear to be too difficult.

**Comment by effective-light** [comment]
Alright, I rebased the branch so you will probably have to reclone.

**Comment by pfps** [comment]
@effective-light Thanks.  I fetched the new version of the PR.

I was looking further into Bluetooth support and I think I know everything that has to be done.  When this PR is in I'll work on Bluetooth support.

File: lib/logitech_receiver/device.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+                    self.wpid = _strhex(device_info[3:5])
+                    self._power_switch = '(' + _('unknown') + ')'
+
+            # the wpid is necessary to properly identify wireless link on/off
+            # notifications also it gets set to None on this object when the
+            # device is unpaired
+            assert self.wpid is not None, 'failed to read wpid: device %d of %s' % (number, receiver)
+
+            for dev in _hid.enumerate({'vendor_id': 0x046d, 'product_id': int(self.receiver.product_id, 16)}):
```
Comment (effective-light) [inline lib/logitech_receiver/device.py:131]: Can you print the `dev.serial` in this loop (and post the output), it seems like unifying devices structure the string differently, we might need to use the `HID_ID` instead if it turns out to be inconsistent. @pfps

Comment (pfps) [reply lib/logitech_receiver/device.py:131]: I added `print("ENUMERATE",dev.serial,dev)`  Here is the result

```
ENUMERATE 1025-93-e8-52-7f DeviceInfo(path='/dev/hidraw1', vendor_id='046d', product_id='c52b', serial='1025-93-e8-52-7f', release=b'1210', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-hidpp-device')
ENUMERATE 1025-93-e8-52-7f DeviceInfo(path='/dev/hidraw1', vendor_id='046d', product_id='c52b', serial='1025-93-e8-52-7f', release=b'1210', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-hidpp-device')
ENUMERATE 4051-60-e4-b9-7d DeviceInfo(path='/dev/hidraw5', vendor_id='046d', product_id='c52b', serial='4051-60-e4-b9-7d', release=b'1210', manufacturer=b'Logitech', product=b'USB Receiver', interface=2, driver='logitech-hidpp-device')
```

Comment (pfps) [reply lib/logitech_receiver/device.py:131]: This might be related to another problem though.   There are two protocols involved with Logitech devices - HID++ and DJ.   The notifications for them are different - to the point that DJ pairing notifications were producing "fake" devices with different WPIDs.  Of course this was all timing-dependent because the processing of the notifications could happen in different orders.  What I did (and this was only very recently added to Solaar) was to ignore DJ pairing notifications.

Comment (effective-light) [reply lib/logitech_receiver/device.py:131]: hm that is strange, are you able to use the "show" command?

Comment (pfps) [reply lib/logitech_receiver/device.py:131]: show works fine, with the same ENUMERATE output.  Note that I have two mice on the receiver, thus the three outputs (one for when looking for the first device, two when looking for the second).

Comment (effective-light) [reply lib/logitech_receiver/device.py:131]: Ah, I haven't touched anything involving notifications yet.

Comment (pfps) [reply lib/logitech_receiver/device.py:131]: It may be that you don't have to worry about the DJ notifications at all.  The recent changes mean that these fake devices don't show up in Solaar any more.

**Comment by pfps** [comment]
With the rebase the only error I'm seeing is when pairing.  The pairing happens but both the CLI and the UI have glitches.

**Comment by pfps** [comment]
Oops, also an error when unpairing the CLI.
```
idefix Solaar> bin/solaar -dd unpair 2
15:58:21,432    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 83B5 030000]
15:58:21,435    DEBUG [MainThread] logitech_receiver.base: (4) => r[11 FF 83B5 031EB692A315065B0000000000000000]
15:58:21,435    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 80B2 000000]
15:58:21,536    DEBUG [MainThread] logitech_receiver.base: (4) => r[20 01 4101 2510040000000000000000]
15:58:21,536    DEBUG [MainThread] logitech_receiver.base: (4) => r[20 02 4101 5140040000000000000000]
15:58:21,536    DEBUG [MainThread] logitech_receiver.base: (4) => r[20 00 4102 0000000000000000000000]
15:58:21,537    DEBUG [MainThread] logitech_receiver.base: (4) => r[10 FF 8F80 B20300]
15:58:21,537    DEBUG [MainThread] logitech_receiver.base: (4) device 0xFF error on request {80B2}: 3 = invalid value
15:58:21,537    DEBUG [MainThread] solaar.cli: [/dev/hidraw0] => <UnifyingReceiver(/dev/hidraw0,4)>
15:58:21,605    DEBUG [MainThread] logitech_receiver.base: (4) <= w[10 FF 83B5 210000]
15:58:21,607    DEBUG [MainThread] logitech_receiver.base: (4) => r[11 FF 83B5 215B0840510402020C00000000000000]
solaar: assertion failed: /home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py line 137
```

**Comment by pfps** [comment]
And an error in `solaar -dd show` when there are no devices.
```

idefix Solaar> bin/solaar show
Unifying Receiver
  Device path  : /dev/hidraw0
  USB id       : 046d:c52b
  Serial       : 1EB692A3
    Firmware   : 12.10.B0032
    Bootloader : 02.14
    Other      : AA.AA
  Has 1 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 2=137
solaar: assertion failed: /home/local/SoftwareDownloads/Solaar/lib/logitech_receiver/device.py line 137
```

**Comment by pfps** [comment]
You can't identify devices by WPID because there might be two devices with the same WPID.

**Comment by pfps** [comment]
@FFY00 This could be a very useful addition to Solaar.  But it looks as if it is breaking because it can't make the connection between a device attached to a receiver and the hidraw node.  Do you know of any way to determine this connection, either given a receiver and a paired device find the hidraw node for the device or given an hidraw node find the receiver and pairing number for the device?  I think I know how to find the receiver from the hidraw node but not anything else above.

**Comment by effective-light** [comment]
@pfps would you happened to know if the serial number is distinct between devices (of the same model)?

**Comment by pfps** [comment]
For devices that have serial numbers, yes, but there may be (older) devices without serial numbers.

**Comment by effective-light** [comment]
@pfps we can fallback to the reciever's handle for those devices

**Comment by pfps** [comment]
Yes, that would be fine.

**Comment by effective-light** [comment]
checkout the latest commit

**Comment by FFY00** [comment]
Yes, it is implemented [here](https://github.com/FFY00/logitechd/blob/08c97d056291d8c880a86b9bc39d3b1e6af4b89f/logitechd/utils.py#L82) see https://github.com/FFY00/logitechd/blob/master/logitechd/__main__.py for an example usage. The code needs to be adapted for Solaar, which could be a problem.

**Comment by pfps** [comment]
@FFY00 I'm not sure that that code is adequate.  It seems to be matching USB manufacturer id and product id.  But that's for receivers - devices are distinguished by WPID and serial number.   But I think @effective-light has the right idea - the serial number can be (usually?) determined as part of the HID_UNIQ field of a python udev device record.

**Comment by pfps** [comment]
@effective-light That's working, but only because I don't have two devices with the same WPID on my receiver.  I think you should try to match the WPID and serial number against the dev serial number.
I forced a mismatch to see whether the code worked if there was no handle and things seemed to be working much better.  There still is the problem with pairing, though.

**Comment by pfps** [comment]
I'm only seeing the pairing problem when the device doesn't have a handle.  When the hidraw node is found things are better.

**Comment by pfps** [comment]
Hmm.  Maybe unpairing and pairing are OK now even if the device doesn't get a handle.  When you get the serial number stuff fixed up I'll check again.

**Comment by effective-light** [comment]
@pfps that is what the code is doing (as of the latest commit), first it reads the device's serial number from the receiver's handle then it tries to match it with the hidraw node that has the same serial number and relevant `wpid`.

**Comment by FFY00** [comment]
Depending on the serial number can come to bite us in the ass I think, but could be explored.

My code searches for devices matching a given list of PIDs, then it uses UDEV to discover the child hidraw nodes associated with the device. See https://github.com/FFY00/logitechd/blob/master/logitechd/utils.py#L82.

**Comment by effective-light** [comment]
I'm probably gonna work on integrating wired devices to `lib/solaar/listener.py` next.

**Comment by pfps** [comment]
Here is a problem with the current way of finding hidraw nodes.
You can have multiple receivers and each of them can think they are paired with a particular device.  Most (but not all) devices will remember which receiver they are paired with.  (There is a particular cracking attack for those devices that do not remember.)  Each pairing generates an hidraw node, which have the same WPID and serial number.
So to find the matching hidraw node you have to get the USB bus and device numbers for the receiver and match them up against the USB bus and device numbers for the device.

**Comment by FFY00** [comment]
What do you mean? The child devices are exported as children in UDEV, we know that for eg. `/dev/hidraw4` is a child of `/dev/hidraw3`, we don't need to do any matching.

**Comment by FFY00** [comment]
The hid-logitech-dj kernel driver has this information, and it exports it to userspace via UDEV.

**Comment by pfps** [comment]
@FFY00 Yes, that would also work.

**Comment by FFY00** [comment]
```
$ ls /sys/module/hid_logitech_dj/drivers/hid:logitech-djreceiver/0003:046D:C53A.008A
0003:046D:405F.008C/  0003:046D:4079.008B/  country  driver@  hidraw/  modalias  power/  report_descriptor  subsystem@  uevent
```

As you can see, `0003:046D:405F.008C` and `0003:046D:4079.008B` are children of `0003:046D:C53A.008A`.

**Comment by pfps** [comment]
@FFY00 Is it possible to get the pairing slot number out of the udev information?  The driver must know it.

**Comment by FFY00** [comment]
Yes, in the physical name. In the last field, `usb-0000:00:14.0-9/input2:1` has ID 1, `usb-0000:00:14.0-9/input2:7` has ID 7 (usually devices only go up to 6, but this is a special read-only device exported in the Powerplay mat/receiver for the built-in led).

**Comment by pfps** [comment]
@FFY00 Excellent!

**Comment by pfps** [comment]
So to find the hidraw node for a pairing slot for the receiver can be done by
looking for the child of the receiver's udev node that has HID_PHYS ending with the slot number (e.g., ...:...:2)
and then looking for that node's child that has SUBSYSTEM hidraw and getting the DEVNAME from that node.  That avoids problems with serial numbers.

**Comment by FFY00** [comment]
That should work, yes.

**Comment by effective-light** [comment]
I can switch to looking for that if it is preferred. @pfps

**Comment by pfps** [comment]
OK, here is code for udev.py that seems to do the trick

```
def find_paired_node(receiver_file, index):
    """Find device paired with receiver"""
    import re
    device =_Devices.from_device_file(_Context(), receiver_file)
    for child in device.find_parent('usb', 'usb_device').children:
        if child.get("HID_PHYS") and re.search(':' + str(index) + '$', child.get("HID_PHYS")):
            for gchild in child.children:
                if gchild.get("SUBSYSTEM") == "hidraw":
                    return gchild.get("DEVNAME")
    return None
```
call from device.py as
```
            self.path = _hid.find_paired_node(receiver.path, number)
            self.handle = _hid.open_path(self.path) if self.path else None
```

The function needs to be added to hidapi/__init__.py

**Comment by pfps** [comment]
@effective-light This way should be better as there are some devices that don't seem to have serial numbers.  This should always work.

**Comment by pfps** [comment]
@effective-light Well it *should* always work but I found a race condition.   Solaar may need to wait for a newly-paired device to show up in udev.  Things still work, as far as I can tell, as communication falls back to the receiver node but it would be nice to retry.  What I did was:
```
            self.path = _hid.find_paired_node(receiver.path, number)
            if self.path is None:
                sleep(1)
                self.path = _hid.find_paired_node(receiver.path, number)
            self.handle = _hid.open_path(self.path) if self.path else None
```
That worked, but there probably is a more elegant way of writing this in Python.

**Comment by pfps** [comment]
A bug fix in receiver.py has created a conflict with this PR.  It's probably a good idea to try to rebase as soon as possible.  I think the correct resolution is to apply the change in PR #920 to the appropriate line in device.py.

**Comment by pfps** [comment]
What's the current status of this PR?

**Comment by effective-light** [comment]
@pfps I want to add at least GUI support for wired devices before we move to testing.

**Comment by effective-light** [comment]
never mind we can move on to testing, I think the GUI additions deserve their own PR.

**Comment by effective-light** [comment]
the only detail of this PR that is somewhat sketchy currently is that I assume all wired devices will use interface 2 for `hidpp` messages (I'm not sure if that is always true).

**Comment by pfps** [comment]
What happens with wired devices?

**Comment by effective-light** [comment]
@pfps it just won't be able to find the device if the device doesn't create a hidraw node for interface 2.

**Comment by pfps** [comment]
Actually, I meant what happens in the GUI.

**Comment by effective-light** [comment]
@pfps All of the GUI stuff relies heavily on the listener classes, so until they are refactored, wired devices won't show up on the GUI whatsoever, it also seems like you would have to implement the `hidpp20` notification scheme since most more recent (non-receiver) devices won't send `hidpp10` notifications.

**Comment by pfps** [comment]
@effective-light Solaar has both hidpp20 and hidpp10 notifications implemented, as far as I know.  There may be some hidpp20 notifications that are missing, but it looks to me that most of the important ones are implemented - battery status, battery voltage, and wireless device status (power).

**Comment by pfps** [comment]
@effective-light So what needs to be done is to generalize ReceiverListener so that it can work for devices with no receiver?

**Comment by barolo** [comment]
From the CLI everything works, pair/unpair, show, config, from the UI too.

**Comment by pfps** [comment]
@barolo Thanks.   It still would be nice to get another tester, but it looks as if these changes are OK.
